### DISCUSSION TITLE
How do I override the base font size?

### DISCUSSION THREAD

**Discussion by shortcircuit3** [tailwindlabs/tailwindcss#1611] [General]
I'd like to adjust the font size system by setting the base font size. How do I set this?

**Comment by ghost** [comment]


**Reply by indr** [reply]
I'm not sure if I mean the same as @milesalex. I'm wondering if there is a way to set the base for the `rem` unit in a responsive manner. That is, say for small screens 1rem = 12px, for large screens 1rem = 14px. 

www.tailwindcss.com actually does set a "base font size" like this:

```css
html {
 font-size:14px
}
@media (min-width:420px) {
 html {
  font-size:16px
 }
}
```

I think the documentation doesn't say anything about setting a "base font size". Ideally we can set these sizes in accordance with the breakpoint system.

**Reply by 18121A05L2** [reply]
hey do we implement this in the global css or in the tailwind config of nextjs ?

**Reply by ghost** [reply]
I decided I would only use tailwind classes in my nextJs 13 demo, and the way I solved the base font was by adding arbitrary values in the className of my HTML tag and targeting screen sizes.

For example, I am using a 15px font size for mobile and 20px for small and bigger screens.

```tsx
import Link from 'next/link'
import 'tailwindcss/tailwind.css'

export default function RootLayout({
  children,
}: {
  children: React.ReactNode
}) {
  return (
    <html
      lang="en"
      className="text-[15px] sm:text-[20px]"
    >
    ...
    <html>
```

**Comment by adamwathan** [comment]
If you just want to change the root font size globally and have everything else scale automatically, do it by changing the font size on the `html` element in your CSS:

```css
html {
  font-size: 20px
}
```

**Reply by shortcircuit3** [reply]
Awesome. Just the answer I was looking for.

**Reply by GoodJob** [reply]
it changes size of all elements of the page :(

**Reply by mzgajner** [reply]
That's pretty much the idea, since everything is based on `rem`s and they're based on the root font size.

**Reply by colorfulfool** [reply]
@adamwathan, is there a way to set the base unit that would change these helpful comments as well? My base unit is to 5px (1rem = 20px), but the comments still assume it's 4px:

<img width="992" alt="Снимок экрана 2021-04-01 в 16 49 14" src="https://user-images.githubusercontent.com/1292290/113304961-3818ba00-930b-11eb-848e-692273a5c8c0.png">

**Reply by datadu-de** [reply]
there should be some way to define it in the theme, I couldnt figure it out yet ...

**Reply by xdivby0** [reply]
Honestly, this is SO what I needed. I found myself ALWAYS setting different heading sizes and paddings for mobile and desktop and that's just bad for readability, writability etc.  This solved almost everything. 

Please add the hint that something like this is helpful in the docs! :D

**Reply by ortonomy** [reply]
This 'works' in that the REM values are all changed -- but the preview available at `/_tailwind` is still broken:

I default to making 1rem = 10px and go from there, but the config UI still reports (and renders the fonts as if it was set to the original REM value)

I really feel like this value should be configurable in the theme ....

![Selection_001](https://user-images.githubusercontent.com/6688676/132954837-0993b619-3b7c-4e55-a488-3f98c1723fa8.png)

**Reply by claudivanfilho** [reply]
Guys, there is a way to solve this! You just need to put an extra property in your VSCode settings.json
```
"tailwindCSS.rootFontSize": 10 // <- your root font size here
```

The result is this
![Screen Shot 2021-11-13 at 9 49 29 AM](https://user-images.githubusercontent.com/4960686/141644433-ac5c918a-fdb0-4c2e-9edb-6ac877409604.png)

**Reply by 18121A05L2** [reply]
> If you just want to change the root font size globally and have everything else scale automatically, do it by changing the font size on the `html` element in your CSS:
> 
> ```css
> html {
>   font-size: 20px
> }
> ```

in which file of tailwind and how we implement it exactly ?

**Reply by maljukan** [reply]
Hah, for all of you like the neat conversion `1rem=10px`:
```CSS
html {
  font-size: 10px
}
```
```HTML
<!-- 12px from here on (as 12px is my base config value)-->
<body class="text-base">
```
With this, you have the conversion and whatever value you put as a base `fontSize{... base: 1.2 ...}` within `tailwind.config.js` will cascade down as the default text `fontSize`

**Comment by usirg** [comment]
Tailwind seems to promote using rem for spacing. And being new to tailwind, I got the impression that this was in the interest of being able to automatically scale dimensions based on rem. But it seems that out of the box, tailwind does not currently support modifying the base font size via utility classes. [Here](https://stackoverflow.com/questions/65232320/responsive-font-sizes-plugin) is a suggestion for how to add this to tailwind - but I haven't tested it yet.

[This](https://elad.medium.com/update-your-font-size-to-rem-units-3f9cba4266c9) article provides a good discussion of the idea of scaling via rem and [this](https://css-tricks.com/rems-ems/) article provides a quite interesting critical discussion of the idea as well as an interesting suggestion for how to deal with some of the pitfalls.

This might be a candidate for where the skilled tailwind folks could come up with a well considered solution, like with the typography plugin. Would love to see that, @adamwathan  :-)

**Reply by datadu-de** [reply]
Having a "official" way of doing this would be awesome!

**Comment by amritk** [comment]
It would be nice to be able to only set the base font-size for tailwind classes as well. I just found out when adding tailwind to an existing project with a base font-size of 10, all of the rem values need to be changed manually. Would be nice if you could change it in the tailwind.config somewhere.

**Reply by hsvikum** [reply]
Faced this exact issue today. Issue made sense but it really caught me off guard. Better to have a official solution.

**Reply by mitul-s** [reply]
Any luck with a simpler solution?

**Reply by oayoub84** [reply]
All what you have to do is extends the spacing values using a function in the config, for example:
```javascript
module.exports = {
  content: ["./src/**/*.{html,ts}"],
  theme: {
    extend: {
      spacing: () => ({
        ...Array.from({ length: 96 }, (_, index) => index * 0.5)
          .filter((i) => i)
          .reduce(
            (acc, i) => ({ ...acc, [i]: `${i / (baseFontSize / 4)}rem` }),
            {}
          ),
      }),
      fontSize: {
        xs: [
          `${(16 * 0.75) / baseFontSize}rem`, /* 12px */
          {
            lineHeight: `${(16 * 1) / baseFontSize}rem` /* 16px */,
          },
        ],
        sm: [
          `${(16 * 0.875) / baseFontSize}rem`, /* 14px */
          {
            lineHeight: `${(16 * 1.25) / baseFontSize}rem` /* 20px */,
          },
        ],
        base: [
          `${(16 * 1) / baseFontSize}rem`, /* 16px */
          {
            lineHeight: `${(16 * 1.5) / baseFontSize}rem` /* 24px */,
          },
        ],
        lg: [
          `${(16 * 1.125) / baseFontSize}rem`, /* 18px */
          {
            lineHeight: `${(16 * 1.75) / baseFontSize}rem` /* 28px */,
          },
        ],
        xl: [
          `${(16 * 1.25) / baseFontSize}rem`, /* 20px */
          {
            lineHeight: `${(16 * 1.75) / baseFontSize}rem` /* 28px */,
          },
        ],
        "2xl": [
          `${(16 * 1.5) / baseFontSize}rem`, /* 24px */
          {
            ineHeight: `${(16 * 2) / baseFontSize}rem` /* 32px */,
          },
        ],
        "3xl": [
          `${(16 * 1.875) / baseFontSize}rem`, /* 30px */
          {
            lineHeight: `${(16 * 2.25) / baseFontSize}rem` /* 36px */,
          },
        ],
        "4xl": [
          `${(16 * 2.25) / baseFontSize}rem`, /* 36px */
          {
            lineHeight: `${(16 * 2.5) / baseFontSize}rem` /* 40px */,
          },
        ],
        "5xl": [
          `${(16 * 3) / baseFontSize}rem`, /* 48px */
          {
            lineHeight: (16 * 1) / baseFontSize,
          },
        ],
        "6xl": [
          `${(16 * 3.75) / baseFontSize}rem`, /* 60px */
          {
            lineHeight: (16 * 1) / baseFontSize,
          },
        ],
        "7xl": [
          `${(16 * 4.5) / baseFontSize}rem`, /* 72px */
          {
            lineHeight: (16 * 1) / baseFontSize,
          },
        ],
        "8xl": [
          `${(16 * 6) / baseFontSize}rem`, /* 96px */
          {
            lineHeight: (16 * 1) / baseFontSize,
          },
        ],
        "9xl": [
          `${(16 * 8) / baseFontSize}rem`, /* 128px */
          {
            lineHeight: (16 * 1) / baseFontSize,
          },
        ],
      },
    },
  },
};

```
Which the baseFontSize is 10 for your example. and adding override fontSize values.

**Comment by dgrahamlong** [comment]
Here's the config conversion for font size, spacing, border-radius, max-width, and line height (assuming base 10px) in case someone needs it.

```
module.exports = {
  theme: {
    extend: {
      lineHeight: {
        3: '1.2rem',
        4: '1.6rem',
        5: '2.0rem',
        6: '2.4rem',
        7: '2.8rem',
        8: '3.2rem',
        9: '3.6rem',
        10: '4.0rem',
      },
      fontSize: {
        xs: ['1.2rem', { lineHeight: '1.6rem' }],
        sm: ['1.4rem', { lineHeight: '2.0rem' }],
        base: ['1.6rem', { lineHeight: '2.4rem' }],
        lg: ['1.8rem', { lineHeight: '2.8rem' }],
        xl: ['2.0rem', { lineHeight: '2.8rem' }],
        '2xl': ['2.4rem', { lineHeight: '3.2rem' }],
        '3xl': ['3.0rem', { lineHeight: '3.6rem' }],
        '4xl': ['3.6rem', { lineHeight: '4.0rem' }],
        '5xl': ['4.8rem', { lineHeight: '1' }],
        '6xl': ['6.0rem', { lineHeight: '1' }],
        '7xl': ['7.2rem', { lineHeight: '1' }],
        '8xl': ['9.6rem', { lineHeight: '1' }],
        '9xl': ['12.8rem', { lineHeight: '1' }],
      },
      spacing: {
        px: '.1rem',
        0.5: '.2rem',
        1: '.4rem',
        1.5: '.6rem',
        2: '.8rem',
        2.5: '1.0rem',
        3: '1.2rem',
        3.5: '1.4rem',
        4: '1.6rem',
        5: '2.0rem',
        6: '2.4rem',
        7: '2.8rem',
        8: '3.2rem',
        9: '3.6rem',
        10: '4.0rem',
        11: '4.4rem',
        12: '4.8rem',
        13.5: '5.4rem',
        14: '5.6rem',
        16: '6.4rem',
        20: '8.0rem',
        24: '9.6rem',
        28: '11.2rem',
        32: '12.8rem',
        36: '14.4rem',
        40: '16.0rem',
        44: '17.6rem',
        48: '19.2rem',
        52: '20.8rem',
        56: '22.4rem',
        60: '24.0rem',
        64: '25.6rem',
        72: '28.8rem',
        80: '32.0rem',
        96: '38.4rem',
        34: '8.5rem',
        68: '27.2rem',
        82.5: '33.0rem',
        90: '36.0rem',
        100: '40.0rem',
        106: '42.4rem',
        200: '80.0rem',
      },
      borderRadius: {
        sm: '.2rem',
        DEFAULT: '.4rem',
        md: '.6rem',
        lg: '.8rem',
        xl: '1.2rem',
        '2xl': '1.6rem',
        '3xl': '2.4rem',
      },
      minWidth: (theme) => ({
        ...theme('spacing'),
      }),
      maxWidth: (theme) => ({
        ...theme('spacing'),
        0: '0rem',
        xs: '32.0rem',
        sm: '38.4rem',
        md: '44.8rem',
        lg: '51.2rem',
        xl: '57.6rem',
        '2xl': '67.2rem',
        '3xl': '76.8rem',
        '4xl': '89.6rem',
        '5xl': '102.4rem',
        '6xl': '115.2rem',
        '7xl': '128.0rem',
      }),
    },
  },
  plugins: [],
}

```

**Reply by danestves** [reply]
Thank you for this! This was that i looking for!

**Reply by meshashub** [reply]
how to change font weight for all the sizes? can it be added somehow into the "fontSize:" ?

**Reply by arun369** [reply]
@earlyaugust2 this is what I was looking for, Thanks a lot. Also if anyone is using VS code is not seeing proper base 10 font sizes in autosuggestion, set this to VS code editor settings

"tailwindCSS.rootFontSize": 10


as suggested by @claudivanfilho above.

**Comment by igdev116** [comment]
`@layer base {
  @font-face {
    font-family: 'Montserrat';
    font-weight: 700;
    font-size: 30px;
  }
}
`
I have tried this but it's not work :(

**Comment by deanmcpherson** [comment]
I wrote a small mod to iterate through the default theme and convert rem to px at a fixed size, runs in my tailwind.config.js

```
const defaultTheme = require('tailwindcss/defaultTheme');

const REM_SIZE = 16;
function checkValue(obj, key, value) {
  if (typeof value === 'string' && value.endsWith('rem')) {
    obj[key] = parseFloat(value) * REM_SIZE + 'px';
  }
}

const mod = obj => {
  if (typeof obj !== 'object') {
    return;
  }
 
  if (Array.isArray(obj)) {
    obj.forEach((value, key) => {
      mod(value);
      checkValue(obj, key, value);
    });
    return;
  }
  Object.keys(obj).forEach(key => {
    const value = obj[key];
    mod(value);
    checkValue(obj, key, value);
  });
};

mod(defaultTheme);

module.exports = {
  content: [
    "./src/pages/**/*.{html,js,tsx,jsx,css}",
    "./src/components/**/*.{html,js,tsx,jsx,css}",
  ],
  theme: {
    extend: {
        ...defaultTheme
    },
},
  plugins: [],
}


```

**Reply by nireno** [reply]
You my savior.
[Here's a modified version](https://gist.github.com/nireno/7669850dffb41e7bede90b9caa5572ca) that allows you to keep  the rem units but just scales the values to suit.
Useful if you want to keep rem units and if for example you're working in the context of a root font size of 11px but you need tailwind to "behave" as if something like w-4 still resolves to something approximating 16px,  you should be able to do `scaleRem(defaultTheme, 11, 16)`.

**Comment by kvedantmahajan** [comment]
This is really bad. We have a npm package coded in tailwindcss. As soon as it got integrated we realised that it broke `rem` styles for package due to portal having different base font size. 
Just wondering, how was this not a consideration?

**Reply by adamwathan** [reply]
You can configure all of the values in your `tailwind.config.js` file to use any units you like. If you need it to be defined in pixels instead of `rem`, just add your own values. It's impossible for the defaults to be perfect for everyone for every situation, that's why the framework is customizable. Please don't blame me for an issue you encountered when building a package without ever testing it in the place it was intended to be used, that sounds like a flaw in your development process, not a flaw in Tailwind.

**Comment by itm-platform** [comment]
It would be great if we could use `em` in addition to `rem` 

Use case: we create a package or control that integrates in a page that we have no control over.

`rem` is only valid if we control the root font size. Whereas if we could set the font size of the parent element and use `em`, we'd have all the benefits of proportional sizing and none of the burdens of using pixels

**Reply by DarkEye123** [reply]
Exactly my thinking when I hit this issue

**Comment by nikushx** [comment]
This is what I love so much about Tailwind, all I need is a basic CSS property and the entire base font can be different. It's so clean.

**Comment by EamonMcKiernanSA** [comment]
Figured I would leave my modification of the above answers here in case anyone needs in the future ^.^

```/** @type {import('tailwindcss').Config} */

/********************************************************************************************
 *
 * Notes for importing extensions from other projects
 * When you import, run the value through the convert function to match the 62.5% reset
 *
 ********************************************************************************************/

const baseFontSize = 10

const convert = (value) => {
  return( ((16 * value) / baseFontSize) + 'rem' );
}

module.exports = {
  content: [
    "./pages/**/*.{js,ts,jsx,tsx}",
    "./components/**/*.{js,ts,jsx,tsx}",
  ],
  theme: {
    extend: {
      spacing: () => ({
        ...Array.from({ length: 96 }, (_, index) => index * 0.5)
            .filter((i) => i)
            .reduce(
                (acc, i) => ({ ...acc, [i]: `${i / (baseFontSize / 4)}rem` }),
                {}
            ),
      }),
      fontSize: {
        xs: [
          `${convert(0.75)}`, /* 12px */
          {
            lineHeight: `${convert(1)}` /* 16px */,
          },
        ],
        sm: [
          `${convert(0.875)}`, /* 14px */
          {
            lineHeight: `${convert(1.25)}` /* 20px */,
          },
        ],
        base: [
          `${convert(1)}`, /* 16px */
          {
            lineHeight: `${convert(1.5)}` /* 24px */,
          },
        ],
        lg: [
          `${convert(1.125)}`, /* 18px */
          {
            lineHeight: `${convert(1.75)}` /* 28px */,
          },
        ],
        xl: [
          `${convert(1.25)}`, /* 20px */
          {
            lineHeight: `${convert(1.75)}` /* 28px */,
          },
        ],
        "2xl": [
          `${convert(1.5)}`, /* 24px */
          {
            lineHeight: `${convert(2)}` /* 32px */,
          },
        ],
        "3xl": [
          `${convert(1.875)}`, /* 30px */
          {
            lineHeight: `${convert(2.25)}` /* 36px */,
          },
        ],
        "4xl": [
          `${convert(2.25)}`, /* 36px */
          {
            lineHeight: `${convert(2.5)}` /* 40px */,
          },
        ],
        "5xl": [
          `${convert(3)}`, /* 48px */
          {
            lineHeight: `${convert(1)}`,
          },
        ],
        "6xl": [
          `${convert(3.75)}`, /* 60px */
          {
            lineHeight: `${convert(1)}`,
          },
        ],
        "7xl": [
          `${convert(4.5)}`, /* 72px */
          {
            lineHeight: `${convert(1)}`,
          },
        ],
        "8xl": [
          `${convert(6)}`, /* 96px */
          {
            lineHeight: `${convert(1)}`,
          },
        ],
        "9xl": [
          `${convert(8)}`, /* 128px */
          {
            lineHeight: `${convert(1)}`,
          },
        ],
      },
      lineHeight: {
        3: `${convert(0.75)}`, /* 12px */
        4: `${convert(1)}`, /* 16px */
        5: `${convert(1.25)}`, /* 20px */
        6: `${convert(1.5)}`, /* 24px */
        7: `${convert(1.75)}`, /* 28px */
        8: `${convert(2)}`, /* 32px */
        9: `${convert(2.25)}`, /* 36px */
        10: `${convert(2.5)}`, /* 40px */
      },
      borderRadius: {
        sm: `${convert(0.125)}`, /* 2px */
        DEFAULT: `${convert(0.25)}`, /* 4px */
        md: `${convert(0.375)}`, /* 6px */
        lg: `${convert(0.5)}`, /* 8px */
        xl: `${convert(0.75)}`, /* 12px */
        '2xl': `${convert(1)}`, /* 16px */
        '3xl': `${convert(1.5)}`, /* 24px */
      },
      minWidth: (theme) => ({
        ...theme('spacing'),
      }),
      maxWidth: (theme) => ({
        ...theme('spacing'),
        0: '0rem',
        xs: `${convert(20)}`, /* 320px */
        sm: `${convert(24)}`, /* 384px */
        md: `${convert(28)}`, /* 448px */
        lg: `${convert(32)}`, /* 512px */
        xl: `${convert(36)}`, /* 576px */
        '2xl': `${convert(42)}`, /* 672px */
        '3xl': `${convert(48)}`, /* 768px */
        '4xl': `${convert(56)}`, /* 896px */
        '5xl': `${convert(64)}`, /* 1024px */
        '6xl': `${convert(72)}`, /* 1152px */
        '7xl': `${convert(80)}`, /* 1280px */
      }),
    },
  },
  plugins: [],
}
```

**Comment by CedericPrivat** [comment]
For anyone that is looking for a way to set the base font size to 10px but preserve the relative proportions of tailwind's utility classes, I recently published a Tailwind CSS plugin 'tailwindcss-base-font-size' that allows you to do just that. You can check it out [here](https://github.com/CedericPrivat/tailwindcss-base-font-size).

Give it a try and let me know, if you have any questions! 👍

**Reply by minimit** [reply]
thanks it works

**Reply by AlexBrasileiro** [reply]
@CedericPrivat thanks a lot! <3

**Comment by TheRealSamChaney** [comment]
I'm using Next.js 14 and to achieve responsive base text size, I used @apply in the html tag block of my globals.css file between @tailwind components and @tailwind utilities like this:

globals.css:
```
...
@tailwind components;
html {
    @apply text-[16pt] sm:text-[20pt];
}
@tailwind utilities;
...
```
This ensures that on large screens down to the small size cutoff the text will be 20pts, but any other size (i.e. below the cutoff) it will be 16. It also allows for this to be overridden by tailwind utilities if needed in places

**Comment by anthonyivol** [comment]
I use this css var workaround to have a fine grained font-size control per element and keeping ratios :
```javascript
import { fontSize } from 'tailwindcss/defaultTheme';
const sizes = Object.keys(fontSize)

// ... config 
fontSize: Object.fromEntries(sizes.map(
  size => [ size, [`calc( var(--font-scale) * ${fontSize[size][0]})`, `calc( var(--font-scale) * ${fontSize[size][1].lineHeight})`] ]
))
// ... config 
```
That way you can use :
```css

@layer base {
  :root{
    --font-scale: 1;
  }
  html {
      font-size: calc( 16px * var(--font-scale) );
  }
}
.my-element-10px-base{
   --font-scale: 0.625
}
.richtext{
  --font-scale: 1.3
}
```
That's also convenient when using multiple fonts which font-size doesn't match
